rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,783 @@
+
+  /**
+    * Node structure (EmptyItem, Leaf and NodePointer)
+    * There is only one type of element in Tree [[Node]] = Vector[Item]
+    */
+  sealed trait Item
+  final case object EmptyItem                                   extends Item
+  final case class Leaf(prefix: ByteVector, value: ByteVector)  extends Item
+  final case class NodePtr(prefix: ByteVector, ptr: ByteVector) extends Item
+
+  type Node = Vector[Item]
+
+  /**
+    * number of items always constant
+    */
+  val numItems = 256
+
+  /**
+    * Empty node consists only of [[EmptyItem]]'s
+    */
+  val emptyNode: Node = (0 until numItems).map(_ => EmptyItem).toVector
+
+  /**
+    * Binary codecs for serializing/deserializing Node in Radix tree
+    * Coding structure for items:
+    *   EmptyItem                   - Empty (not encode)
+
+    *   Leaf(prefix,value)    -> [item number] [second byte] [prefix0]..[prefixM] [value0]..[value31]
+    *                               where is: [second byte] -> bit7 = 0 (Leaf identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    *   NodePtr(prefix,ptr)   -> [item number] [second byte] [prefix0]..[prefixM] [ptr0]..[ptr31]
+    *                               where is: [second byte] -> bit7 = 1 (NodePtr identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    */
+  object codecs {
+    private val defSize = 32
+
+    def encode(node: Node): ByteVector = {
+      //calculating size of buffer
+      val sizeNode = node.size
+
+      @tailrec
+      def loopSize(index: Int, size: Int): Int =
+        if (sizeNode > index)
+          node(index) match {
+            case EmptyItem => loopSize(index + 1, size)
+
+            case Leaf(leafPrefix, value) =>
+              val sizePrefix = leafPrefix.size.toInt
+              assert(sizePrefix <= 127, "error during serialization (size of prefix more than 127)")
+              assert(
+                value.size == defSize,
+                "error during serialization (size of leafValue not equal 32)"
+              )
+              loopSize(index + 1, size + 2 + sizePrefix + defSize)
+
+            case NodePtr(ptrPrefix, ptr) =>
+              val sizePrefix = ptrPrefix.size.toInt
+              assert(sizePrefix <= 127, "error during serialization (size of prefix more than 127)")
+              assert(
+                ptr.size == defSize,
+                "error during serialization (size of ptrPrefix not equal 32)"
+              )
+              loopSize(index + 1, size + 2 + sizePrefix + defSize)
+          } else size
+
+      val size = loopSize(0, 0)
+      //put in the buffer
+
+      val arr = new Array[Byte](size)
+      @tailrec
+      def loopFill(numItem: Int, idx0: Int): Array[Byte] =
+        if (idx0 == size) arr
+        else {
+          node(numItem) match {
+            case EmptyItem => loopFill(numItem + 1, idx0)
```

Comment:
I am very curious that if we encode nothing in `EmptyItem`. How can you encode `Vector(NodePtr(), EmptyItem, EmptyItem, Leaf())` and decode it back to `Vector(NodePtr(), EmptyItem, EmptyItem, Leaf())` correctly?

Reply:
We may not encode EmptyItem because we store [item number] for all Leafs and NodePtrs. A node always contains 256 items. During decoding all unlisted nodes to defining as EmptyItems.
I will add this info to the codec description.

Reply:
Ok. Some actually the order of `items` inside the `node` is not important, isn't it? But I saw a lot of usage of `node(index)` which the index is kind of breaking something. I think I should miss something in the middle.

Reply:
There are two views of a node:
1) serialized data  (type ByteVector) (used only for saving in RadixStore) - here we don't save EmtyNode and save the number of Items. In this case, we don't need to sort items by numbers. But this is still done during the coding process.
2) Decoded nodes (Vector[Item]) in the form of vectors with which all the work goes. Sorting is critical here.

Probably you think that when serializing data, we could not save numbers, but allocate one bit to store empty ones? I researched different approaches to serialization and the one that is implemented is optimal for our task in my opinion. 
Most often, we have sparse nodes, that is, there are many empty and few not empty ones. Keeping empty nodes on serialization slowed down performance.